pkg/pool-weighted/contracts/lbp/LiquidityBootstrappingPool.sol

```diff
@@ -123,166 +61,96 @@ contract LiquidityBootstrappingPool is IMinimalSwapInfoPool, NewBasePool {
+                request.amount
+            );
 
-    /**
-     * @notice Return whether swaps are enabled or not for the given pool.
-     */
-    function getSwapEnabled() public view returns (bool) {
-        return _poolState.decodeBool(_SWAP_ENABLED_OFFSET);
-    }
+            // amountOut tokens are exiting the Pool, so we round down.
+            return _downscaleDown(amountOut, scalingFactorTokenOut);
+        } else {
+            // All token amounts are upscaled.
+            request.amount = _upscale(request.amount, scalingFactorTokenOut);
 
-    /**
-     * @notice Return the current value of the swap fee percentage.
-     * @dev This is stored separately, as there is no more room in `_poolState`.
-     */
-    function getSwapFeePercentage() public view virtual override returns (uint256) {
-        return _swapFeePercentage;
-    }
+            uint256 amountIn = WeightedMath._calcInGivenOut(
+                balanceTokenIn,
+                _getNormalizedWeight(request.tokenIn),
+                balanceTokenOut,
+                _getNormalizedWeight(request.tokenOut),
+                request.amount
+            );
 
-    /**
-     * @notice Return the current token weights.
-     */
-    function getNormalizedWeights() external view returns (uint256[] memory) {
-        return _getNormalizedWeights();
-    }
+            // amountIn tokens are entering the Pool, so we round up.
+            amountIn = _downscaleUp(amountIn, scalingFactorTokenIn);
 
-    /**
-     * @dev Return start time, end time, and endWeights as an array.
-     * Current weights should be retrieved via `getNormalizedWeights()`.
-     */
-    function getGradualWeightUpdateParams()
-        external
-        view
-        returns (
-            uint256 startTime,
-            uint256 endTime,
-            uint256[] memory endWeights
-        )
-    {
-        // Load current pool state from storage
-        bytes32 poolState = _poolState;
-
-        startTime = poolState.decodeUint(_START_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
-        endTime = poolState.decodeUint(_END_TIME_OFFSET, _TIMESTAMP_BIT_LENGTH);
-        uint256 totalTokens = _getTotalTokens();
-        endWeights = new uint256[](totalTokens);
-
-        for (uint256 i = 0; i < totalTokens; i++) {
-            endWeights[i] = poolState
-                .decodeUint(_END_WEIGHT_OFFSET + i * _END_WEIGHT_BIT_LENGTH, _END_WEIGHT_BIT_LENGTH)
-                .decompress(_END_WEIGHT_BIT_LENGTH);
+            // Fees are added after scaling happens, to reduce the complexity of the rounding direction analysis.
+            return _addSwapFeeAmount(amountIn);
         }
     }
 
     /**
-     * @notice Pause/unpause trading.
+     * @dev Adds swap fee amount to `amount`, returning a higher value.
      */
-    function setSwapEnabled(bool swapEnabled) external authenticate whenNotPaused {
-        _setSwapEnabled(swapEnabled);
+    function _addSwapFeeAmount(uint256 amount) internal view returns (uint256) {
```

Comment:
Seems we could also inline _addSwapFeeAmount and _subtractSwapFeeAmount, since they're only used here now.

Reply:
Yep, we also need to update the swap/join/exit hooks to pass an explicit `poolState` parameter around which should be handled at the same time imo. My thinking is that this falls out of scope for this PR and we can clean this up later.